Develop

haenghaeng opened the pull request:
## ⭐Key Changes
1. Problem에 있던 일부 프로퍼티를 분리하여 ProblemDetail 클래스를 만들었습니다.
2. Problem과 ProblemDetail 사이의 의존관계를 만들었습니다.
3. ProblemRepository에 문제의 제목으로 Problem만 검색하는 기능과 Detail을 같이 검색하는 기능을 추가하였습니다.

23.10.14 추가내용
1. Problem과 ProblemDetail간의 관계를 단방향으로 바꾸었습니다.
2. registeredTime은 중복되는 프로퍼티라서 제거하였습니다.
3. Repository에서 find method의 방식을 문제 제목이 아니라 id를 기준으로 찾게 변경하였습니다.
4. 통과 비율 계산에서 divisionByZero가 나지 않게 if문을 추가하였습니다.

## 📌 issue
close #11 

## To Reviewers

1.
findWithProblemDetailById 메소드의 함수명을 findByIdWithProblemDetail로 변경하니 오류가 발생합니다.
명명법에 위배되는것 같습니다. Reference를 계속 살펴보고 있는 중입니다.

=> findWithProblemDetailById은 쿼리를 살펴보니 join을 하지 않네요.. 쿼리 공부를 더 하겠습니다...

2. 
@Enumerated(EnumType.STRING)
@ElementCollection(fetch = FetchType.EAGER)
private Set<EAvailableLanguage> availableLanguage;
테스트를 계속 돌려보고 설정을 바꾸어 보았지만 계속해서 LazyInitializationException이 발생합니다.
@Transactional를 추가하여 테스트를 진행해도 같은 결과가 나옵니다. 이 부분은 나중에 알아보고 다시 수정하겠습니다.

=> 위의 함수를 그대로 사용해 테스트를 진행해서 같은 결과가 나온 것 같습니다... 쿼리 공부를 열심히 하겠습니다...


## Next move
다양한 검색 기능 추가하기(ex. 난이도별, 합격비율 오름차순/내림차순)

haenghaeng committed:
Merge branch 'develop' into main

haenghaeng committed:
Merge remote-tracking branch 'origin/main' into main

haenghaeng committed:
Merge branch 'develop' into main

haenghaeng committed:
feat:ProblemDetail 추가

Problem에 있는 일부 프로퍼티를 ProblemDetail로 옮겼습니다. 또한 새로운 프로퍼티를 추가하였습니다

haenghaeng committed:
remove:ProblemFactory,ProblemRequest

나중에 다시 구현할 예정입니다.

haenghaeng committed:
fix:Problem

fetchLazy설정이 빠져 추가합니다.
index는 @NotBlank이지만 아직 생성 로직을 적어 놓지 않아 오류가 발생합니다. 따라서 임시로 주석 처리 합니다.

haenghaeng committed:
fix:ProblemDetail

주석을 추가하고 JoinColumn의 잘못된 부분을 수정하였습니다. 또한 Limit에 적혀있던 상수를 제거하고 생성자에 넣었습니다.

haenghaeng committed:
feat:ProblemRepository

title을 기준으로 검색하는 기능을 추가하였습니다. 하나는 fetch join을 사용하여 problem, detail을 둘 다 검색합니다. 다른 하나는 problem만 검색합니다.

haenghaeng committed:
Merge branch 'develop' into main

DebugMind commented on poj/src/main/java/com/poj/entity/problem/Problem.java:
```diff
@@ -25,44 +23,36 @@ public class Problem extends BaseEntity {
-    @NotBlank
-    @Column(unique = true)
-    private Long index; // 번호.
+    @OneToOne(mappedBy = "problem", fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
```
ProblemDetail 은 Problem 에 종속되는 Entity 입니다.
따라서 (cascade = CascadeType.ALL, orphanRemoval = true) 이렇게 설정해주는 것이 좋아보입니다.
cascade = CascadeType.ALL 은 부모 앤티티(Problem) 의 데이터베이스 작업(저장,업데이트,삭제) 이 자식 엔티티(ProblemDetail) 에도 적용되도록 하는 것입니다. 
orphanRemoval = true 은 부모 엔티티에서 자식 엔티티의 참조가 제거될 때 자식 엔티티가 자동으로 삭제되도록 하는 것입니다.
위 두 옵션을 적용하면 Problem 과 ProblemDetail 을 동일한 생명주기를 가지게 되어 무결성이 지켜질 수 있습니다 :)
검색이나 gpt 에 물어보면 더 상세한 정보를 알 수 있으니 한 번 공부해 보시는걸 권장 드립니다 👍

DebugMind commented on poj/src/main/java/com/poj/entity/problem/Problem.java:
```diff
@@ -25,44 +23,36 @@ public class Problem extends BaseEntity {
-
-    private EProblemDifficulty difficulty; // 난이도
-    private List<EAvailableLanguage> availableLanguage; // 문제에서 사용 가능한 언어 리스트(리스트??)
+     private EProblemDifficulty difficulty; // 난이도
```
Enum class 를 그대로 db에 저장하게 되면 db에는 Enum 의 인덱스 (0,1,2 ... ) 로 저장하게 됩니다. 
이렇게 되면 유지 보수 시에 혼동을 가져올 수 있으므로 
```@Enumerated(EnumType.STRING)``` 으로 세팅하는 것을 권장합니다 👍  
이렇게 하면 Enum 의 문자열로 db에 저장됩니다 :)

DebugMind commented on poj/src/main/java/com/poj/entity/problem/Problem.java:
```diff
@@ -25,44 +23,36 @@ public class Problem extends BaseEntity {
-    private EProblemDifficulty difficulty; // 난이도
-    private List<EAvailableLanguage> availableLanguage; // 문제에서 사용 가능한 언어 리스트(리스트??)
+     private EProblemDifficulty difficulty; // 난이도
+     private Set<EAvailableLanguage> availableLanguage; // 문제에서 사용 가능한 언어 집합
```
db 에는 원칙적으로 하나의 칼럼에는 하나의 데이터만 저장할 수 있습니다. 
따라서 Set 과 같이 여러개의 엔티티를 저장하려면 일대다 매핑을 이용해줘야합니다. (일대다은 db 설계 시 반드시 알아야하는 중요한 개념으로 꼭 알아보시는 걸 권장드립니다 👍 ) 
따라서 ```ElementCollection``` 나 ```OneToMany``` 등 의 사용이 필요해보입니다.

DebugMind commented on poj/src/main/java/com/poj/entity/problem/Problem.java:
```diff
@@ -25,44 +23,36 @@ public class Problem extends BaseEntity {
+     private EProblemDifficulty difficulty; // 난이도
+     private Set<EAvailableLanguage> availableLanguage; // 문제에서 사용 가능한 언어 집합
 
     private Long registeredTime; // 문제를 등록한 시간
```
이 엔티티가 생성되면 createdAt 이라는 프로퍼티가 자동으로 생성됩니다 (BaseEntity 에 있습니다 :)) 
따라서 registeredTime 굳이 필요 없는 프로퍼티라고 생각됩니다 👍

DebugMind commented on poj/src/main/java/com/poj/entity/problem/ProblemDetail.java:
```diff
@@ -0,0 +1,60 @@
+
+    @OneToOne
+    @JoinColumn(name = "problem_id")
+    private Problem problem;
```
Problem 과 ProblemDetail은 현재 양방향 매핑 중인데 ProblemDetail 에서 Problem 을 가져와야하는 경우는 보통 잘 발생하지 않습니다. 따라서 Problem 에서만 ProblemDetail 을 참조하는 단방향 매핑이 더 좋아보입니다. 👍

DebugMind commented on poj/src/main/java/com/poj/repository/problem/ProblemRepository.java:
```diff
@@ -1,11 +1,23 @@
+    Optional<Problem> findByTitle(String title); // problem only
+
+    @EntityGraph(attributePaths = {"problemDetail"})
+    Optional<Problem> findWithProblemDetailByTitle(String title); // fetch join
```
메서드 변수 명 패턴이 findByTitleWithProblemDetail 이런식으로 By 가 더 앞에 나오게 짓고 있어서
일관성을 위해서 이렇게 수정해주시면 감사하겠습니다!! :)

DebugMind replied:
추가적으로 앤티티를 조회할 때는 앤티티의 고유 키 (problem_id) 로 조회하는 것이 일반적입니다. pk로 조회할 경우 시간복잡도가 O(1) 로 매우 효율적이기 때문이죠 ㅎㅎ

Hosung-Ahn commented:
추가적으로 Problem Entity 의 ``` getPassRatio `` `의 함수에서 submitNumber 이 0일 경우 divisionByZero 에러가 발생할 수 있을 것 같습니다. 여기에 대해서 예외처리 해주시면 감사하겠습니다. 잘못된 코드 짜드려서 죄송합니다 ㅜ,ㅜ

haenghaeng committed:
refactor:ProblemDetail

양방향 참조에서 단방향 참조로 변경합니다.

haenghaeng committed:
refactor:Problem

Detail과의 양방향 참조에서 단방향 참조로 바꾸었습니다.
문제 등록 시간은 중복되는 프로퍼티이기에 제거하였습니다.
Enumerator 타입이 String으로 저장되게 바꾸었습니다.

haenghaeng committed:
refactor:ProblemRepository

찾는 방법을 title이 아니라 ID방식으로 변경하였습니다.

haenghaeng committed:
fix:Problem

통과 비율 계산에서 divisionByZero 에러가 나지 않게 예외처리를 하였습니다.

DebugMind committed:
fix : findByIdWithProblemDetail 구현 - querydsl

DebugMind committed:
test : Problem 과 ProblemDetail 의 생성 로직 테스트

Problem 과 ProblemDetail 이 fetch lazy 전략으로 작동하는지 테스트 하였습니다.

Hosung-Ahn commented:
LazyInitializationException 은 fetch lazy 로 설정한 객체를 영속성 컨텍스트의 주기(주로 transactional 주기) 가 종료된 후 참조하려고 할 때 발생합니다. 일단 Problem 과 ProblemDetail의 생성작업에 대한 테스트 코드를 만들었습니다. 👍 
LazyInitializationException 이 발생하는 상황에 대해서 테스트 코드에 작성해 주시면 한번 더 살펴보겠습니다 :)
